biz.aQute.tester.junit-platform/src/aQute/tester/junit/platform/Activator.java

```diff
@@ -209,6 +210,16 @@ public void executionSkipped(TestIdentifier testIdentifier, String reason) {
 				message("", "TEST %s <<< SKIPPED", testName(testIdentifier));
 			}
 		});
+		ServiceTracker<TestExecutionListener, TestExecutionListener> track = new ServiceTracker<>(context,
+			TestExecutionListener.class, null);
+		track.open(true);
+		Optional.ofNullable(track.getServices())
+			.ifPresent(services -> Stream.of(services)
+				.map(TestExecutionListener.class::cast)
+				.forEach(listener -> {
+					listenerList.add(listener);
+				}));
+		track.close();
```

Comment:
This is problematic. When you close the tracker, you release the services. Which means you should no longer use the service objects. But you clearly keep using the service objects here. A bundle providing the service could use a ServiceFactory and disable the object in someway when the service is released. It is never good practice to use a service object after you release the service.

A tracker like this should be created and opened in the bundle's activator start method and closed in the stop method.

Reply:
Thanks for the clarification!

I thought about opening it in the activator start, but in the end I opted for opening it just before the test run starts. The reason for this is because I would like at some time to add the option of supplying a filter to the tracker constructor so that you can control which listeners are included in the test run on a per-run basis. If ever we get the Gogo commandlet working, then in a continuous testing scenario it might be nice to be able to dynamically change that filter and have it take effect straight away without having to restart the test bundle (restarting the test bundle will cause the framework to exit if it has been launched on the main thread). See if what I've done is ok and if not I'll change it.